src/tokenserver/db/models.rs

```diff
@@ -264,6 +265,171 @@ impl TokenserverDb {
             .map_err(Into::into)
     }
 
+    fn get_users_sync(&self, params: params::GetUsers) -> DbResult<results::GetUsers> {
+        const QUERY: &str = r#"
+                     SELECT uid, nodes.node, generation, keys_changed_at, client_state, created_at,
+                            replaced_at
+                       FROM users
+            LEFT OUTER JOIN nodes ON users.nodeid = nodes.id
+                      WHERE email = ?
+                        AND users.service = ?
+                   ORDER BY created_at DESC, uid DESC
+                      LIMIT 20
+        "#;
+
+        diesel::sql_query(QUERY)
+            .bind::<Text, _>(&params.email)
+            .bind::<Integer, _>(params.service_id)
+            .load::<results::GetRawUser>(&self.inner.conn)
+            .map_err(Into::into)
+    }
+
+    /// Gets the user with the given email and service ID, or if one doesn't exist, allocates a new
+    /// user.
+    fn get_or_create_user_sync(
+        &self,
+        params: params::GetOrCreateUser,
+    ) -> DbResult<results::GetOrCreateUser> {
+        let mut raw_users = self.get_users_sync(params::GetUsers {
```

Comment:
I couldn't figure out how to have an async `get_or_create_users` method that runs on the blocking threadpool, since defining that method on the trait would result in the the method body running on an actix worker thread (though the database calls would still run on the blocking threadpool). This seemed preferable given those two options, but this method as it's defined will block a thread for quite a long time, since there are no `await`s. Definitely open to suggestions here

Reply:
Hashing a few things out, so my apologies in advance.
There are two blocking operations in this thread, both of which potentially block the Actix event loop. `get_users_sync`, and `allocate_user_sync`, both of which use the `dbpool` which runs on it's own system thread. `allocate_user_sync` does three operations, while `get_users_sync` just does one GET. If we wanted to free up the actix event loop during those db calls, we could probably wrap those in futures. 

We might want to just put some profile metrics on `allocate_user_sync` to see what sort of delays it introduces before we optimize things.

Reply:
It could potentially be async, since it no longer has any direct calls to diesel now. From an async block it'd call the async versions of the individual db calls it makes -- it looks like it has 3 of them -- each would up separately/individually called via the thread pool.

I'm not terribly concerned about the way it is currently but that's always an option.